backend/core/src/Core.API/ResponseHandling/SignatureVerificationMiddleware.cs

```diff
@@ -0,0 +1,184 @@
+                {
+                    await _next(context); // Signature is valid, continue with the request
+                }
+                else
+                {
+                    // Check the previous timestamp (30 seconds ago)
+                    long previousTimestamp = currentTimestamp - 30;
+
+                    string? updatedPayload = null;
+
+                    switch (method)
+                    {
+                        case "POST":
+                        case "PUT":
+                            {
+                                updatedPayload = JsonConvert.SerializeObject(new
+                                {
+                                    publicKey = publicKeyHeader,
+                                    timestamp = previousTimestamp,
+                                    postPayload
+                                });
+                                break;
+                            }
+
+                        default:
+                            updatedPayload = JsonConvert.SerializeObject(new
+                            {
+                                publicKey = publicKeyHeader,
+                                timestamp = previousTimestamp
+                            });
+                            break;
+                    }
+
+                    payloadBytes = Encoding.UTF8.GetBytes(updatedPayload);
+
+                    // Verify the signature again with the previous timestamp
+                    if (VerifySignature(publicKeyHeader, payloadBytes, signatureBytes))
+                    {
+                        await _next(context); // Signature is valid, continue with the request
+                    }
+                    else
+                    {
+                        _logger.LogError("Invalid signature");
+                        var customErrors = new CustomErrors(new CustomError("Forbidden", "Invalid signature", "x-signature"));
+                        await WriteCustomErrors(context.Response, customErrors, (int)HttpStatusCode.Forbidden);
+                    }
+                }
+            }
+            catch (Exception ex)
+            {
+                _logger.LogError("Unknown exception thrown: {message}", ex.Message);
+                throw new Exception("Unknown exception thrown: {message}", ex);
+            }
+        }
+
+        private static async Task WriteCustomErrors(HttpResponse httpResponse, CustomErrors customErrors, int statusCode)
+        {
+            httpResponse.StatusCode = statusCode;
+            httpResponse.ContentType = "application/json";
+
+            var response = CustomErrorsResponse.FromCustomErrors(customErrors);
+            var json = System.Text.Json.JsonSerializer.Serialize(response);
+            await httpResponse.WriteAsync(json);
+        }
+
+        private static bool VerifySignature(string publicKey, byte[] payload, byte[] signature)
+        {
+            publicKey = publicKey.Replace("\\n", "\n");
```

Comment:
Added this, as Frontend does JSON.stringify before adding it in the header
`config.headers["x-public-key"] = JSON.stringify(pubKeyFromStore);`

Reply:
I think it'd be better to strip \\n \r\n and other such things completely (on both ends)